### ISSUE TITLE
Apache HttpClient and Jackson Versions are too old

### ISSUE THREAD

**Issue by MingChenSlot** [elastic/elasticsearch-hadoop#827]
We are using EsBolt in our project. However, our projects depend on apache httpcomponent 4.x and jackson 2.x, and when initializing EsBolt, it calls Hadoop StringUtils (https://github.com/elastic/elasticsearch-hadoop/blob/5.x/mr/src/main/java/org/elasticsearch/hadoop/util/StringUtils.java), which depends on apache httpclient 3.x and jackson 1.x, and we got NoClassFoundException.  Here are some errors,

java.lang.NoClassDefFoundError: org/apache/commons/httpclient/URIException
    at org.elasticsearch.hadoop.cfg.Settings.getProperty(Settings.java:536) 
    at org.elasticsearch.storm.cfg.StormSettings.getStormTickTupleFlush(StormSettings.java:46) 
    at org.elasticsearch.storm.EsBolt.prepare(EsBolt.java:83)

java.lang.NoClassDefFoundError: org/codehaus/jackson/map/ObjectMapper
    at org.elasticsearch.hadoop.rest.RestClient.<init>(RestClient.java:78) 
    at org.elasticsearch.hadoop.rest.InitializationUtils.discoverEsVersion(InitializationUtils.java:189) 
    at org.elasticsearch.hadoop.rest.RestService.createWriter(RestService.java:542) 
    at org.elasticsearch.storm.EsBolt.prepare(EsBolt.java:103)

Do you guys have any plan to upgrade the version of apache httpclient and jackson since current versions are pretty old? 

Thanks a lot.

**Comment by jbaiera** [comment]
@MingChenSlot Since much of the Hadoop ecosystem requires you to install third party libraries some how (either by libjars in Hadoop MR, or explicit installation/loading in Pig and Hive) we have designed ES-Hadoop to use as much of the available libraries that would be already present in Hadoop once installed. The versions of Jackson and commons HTTP are aligned with the versions used in Hadoop so that we can avoid jar versioning issues and to minimize how much setup is required by the end user for the library.

**Comment by jnioche** [comment]
Having a dependency on httpclient in StringUtils just to handle URIs is a bit heavy handed IMHO. Especially given that the authors  recommend using the standard JRE classes instead (see http://marc.info/?l=httpclient-users&m=125425095705062&w=2) and have removed it from recent versions.

I use Apache Storm without Hadoop and would rather avoid having to add yet another dependency which might conflict with the versions of httpclient that I already use.

BTW since there is an assumption that elasticsearch-storm should have a dependency on httpclient, why isn't that made explicit in the Maven dependencies?

**Comment by jbaiera** [comment]
@jnioche In 5.3.3 we moved the URI and URL encoding and utility functions out of StringUtil into its own utility class (HttpEncodingTools). We've also minimized how much we depend on the commons encoding logic, and deprecated the parts that still have a hard dependency on it. As for the maven dependencies, the storm project is normally included in with the rest of the projects in the final jar. My guess is that the dependency was left off as an oversight. 

I'm going to close this issue for the time being and open an issue to address the missing storm dependency.
